File: apps/client/local_plugins/vendors/ast_polychat/android/src/main/kotlin/com/aiagent/ast_polychat/AstPolychatService.kt
package com.aiagent.ast_polychat

import android.content.Context
import android.util.Log
import com.aiagent.plugin_interface.AstCallback
import com.aiagent.plugin_interface.AstRole
import com.aiagent.plugin_interface.NativeAstService
import com.aiagent.plugin_interface.VoitransWebRtcSession
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import org.json.JSONObject
import java.util.concurrent.ThreadLocalRandom

/**
 * PolyChat 平台 AST 服务实现（WebRTC 传输）
 *
 * DataChannel 协议（`trans_original` / `trans_translated`）携带 `done` 标志：
 *   - `done=false` 中间态（累计快照）→ recognizing
 *   - `done=true`  本段定稿       → recognized + recognitionDone(role)
 *
 * 当 source 与 translated 两个角色都 done 后派发 `recognitionEnd`。
 * 用户重新说话（`user_speaking`）会强制 close + endRound 上一回合。
 */
class AstPolychatService(private val context: Context) : NativeAstService {

    companion object {
        private const val TAG = "AstPolychat"
    }

    private lateinit var session: VoitransWebRtcSession
    private var callback: AstCallback? = null
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    // ── Recognition round state ──
    private var currentRequestId: String? = null
    private var sourceRoleOpen = false
    private var translatedRoleOpen = false

    override fun initialize(configJson: String, context: Context) {
        val cfg = JSONObject(configJson)
        VoitransWebRtcSession.warmupHttp(cfg.getString("baseUrl"))
        session = VoitransWebRtcSession(context)
        session.initialize(
            baseUrl = cfg.getString("baseUrl"),
            appId = cfg.getString("appId"),
            appSecret = cfg.getString("appSecret"),
            agentId = cfg.getString("agentId"),
        )
        Log.d(TAG, "Initialized: agentId=${cfg.getString("agentId")}")
    }

    override fun connect(callback: AstCallback) {
        this.callback = callback
        resetRoundState()
        session.connect(object : VoitransWebRtcSession.EventListener {
            override fun onConnected() {
                Log.d(TAG, "Connected")
                callback.onConnected()
            }

            override fun onMessage(json: JSONObject) {
                handleDataChannelMessage(json)
            }

            override fun onDisconnected() {
                Log.d(TAG, "Disconnected")
                forceEndRound()
                callback.onDisconnected()
            }

            override fun onError(code: String, message: String) {
                Log.e(TAG, "Error: [$code] $message")
                callback.onError(code, message)
            }
        })
    }

    override fun startAudio() {
        session.startAudio()
    }

    override fun stopAudio() {
        session.stopAudio()
    }

    override fun interrupt() {
        // WebRTC 模式下打断靠服务端 VAD 检测
    }

    override fun release() {
        forceEndRound()
        scope.cancel()
        session.release()
        callback = null
    }

    // ── DataChannel 事件映射 ──

    private fun handleDataChannelMessage(json: JSONObject) {
        val cb = callback ?: return
        val type = json.optString("type", "")

        when (type) {
            "user_speaking" -> {
                beginRound(cb, force = true)
                openRole(cb, AstRole.SOURCE)
            }

            "trans_original" -> {
                beginRound(cb)
                openRole(cb, AstRole.SOURCE)
                val text = json.optString("text", "")
                val done = json.optBoolean("done", false)
                if (text.isNotEmpty()) {
                    val rid = currentRequestId ?: return
                    if (done) {
                        cb.onRecognized(AstRole.SOURCE, rid, text)
                        closeRole(cb, AstRole.SOURCE)
                        maybeEndRound(cb)
                    } else {
                        cb.onRecognizing(AstRole.SOURCE, rid, text)
                    }
                }
            }

            "trans_translated" -> {
                beginRound(cb)
                openRole(cb, AstRole.TRANSLATED)
                val text = json.optString("text", "")
                val done = json.optBoolean("done", false)
                if (text.isNotEmpty()) {
                    val rid = currentRequestId ?: return
                    if (done) {
                        cb.onRecognized(AstRole.TRANSLATED, rid, text)
                        closeRole(cb, AstRole.TRANSLATED)
                        maybeEndRound(cb)
                    } else {
                        cb.onRecognizing(AstRole.TRANSLATED, rid, text)
                    }
                }
            }

            "error" -> {
                val message = json.optString("message", "Unknown error")
                val fatal = json.optBoolean("fatal", false)
                if (fatal) {
                    cb.onError("ast.fatal", message)
                } else {
                    cb.onRecognitionError(currentRequestId, null, "ast.error", message)
                }
            }

            "session_state",
            "mcp_tool_call",
            "mcp_tool_result",
            "disconnect_warning",
            "user_transcription",
            "bot_response_start",
            "bot_response",
            "ai_speaking",
            "ai_stopped",
            "ai_response_done" -> {
                // Not surfaced through AstCallback.
            }

            else -> {
                Log.d(TAG, "Unknown event: $type")
            }
        }
    }

    // ── Recognition round state machine ──

    private fun beginRound(cb: AstCallback, force: Boolean = false) {
        if (currentRequestId != null) {
            if (!force) return
            if (sourceRoleOpen) closeRole(cb, AstRole.SOURCE)
            if (translatedRoleOpen) closeRole(cb, AstRole.TRANSLATED)
            endRound(cb)
        }
        currentRequestId = newRequestId()
    }

    private fun openRole(cb: AstCallback, role: AstRole) {
        val rid = currentRequestId ?: return
        when (role) {
            AstRole.SOURCE -> {
                if (!sourceRoleOpen) {
                    sourceRoleOpen = true
                    cb.onRecognitionStart(role, rid)
                }
            }
            AstRole.TRANSLATED -> {
                if (!translatedRoleOpen) {
                    translatedRoleOpen = true
                    cb.onRecognitionStart(role, rid)
                }
            }
        }
    }

    private fun closeRole(cb: AstCallback, role: AstRole) {
        val rid = currentRequestId ?: return
        when (role) {
            AstRole.SOURCE -> {
                if (sourceRoleOpen) {
                    sourceRoleOpen = false
                    cb.onRecognitionDone(role, rid)
                }
            }
            AstRole.TRANSLATED -> {
                if (translatedRoleOpen) {
                    translatedRoleOpen = false
                    cb.onRecognitionDone(role, rid)
                }
            }
        }
    }

    private fun maybeEndRound(cb: AstCallback) {
        if (sourceRoleOpen || translatedRoleOpen) return
        if (currentRequestId == null) return
        endRound(cb)
    }

    private fun endRound(cb: AstCallback) {
        val rid = currentRequestId ?: return
        cb.onRecognitionEnd(rid)
        resetRoundState()
    }

    private fun forceEndRound() {
        val cb = callback ?: run { resetRoundState(); return }
        if (currentRequestId == null) return
        if (sourceRoleOpen) closeRole(cb, AstRole.SOURCE)
        if (translatedRoleOpen) closeRole(cb, AstRole.TRANSLATED)
        endRound(cb)
    }

    private fun resetRoundState() {
        currentRequestId = null
        sourceRoleOpen = false
        translatedRoleOpen = false
    }

    private fun newRequestId(): String {
        val ms = System.currentTimeMillis()
        val rand = ThreadLocalRandom.current().nextInt(1 shl 30)
            .toString(36).padStart(6, '0')
        return "ast_polychat_${ms}_$rand"
    }
}

File: apps/client/local_plugins/vendors/ast_polychat/ios/Classes/AstPolychatService.swift
import Foundation
import ai_plugin_interface
import os.log

/// PolyChat speech translation service (WebRTC transport).
///
/// Ports the Android `AstPolychatService`. Maps DataChannel
/// `trans_original` / `trans_translated` frames onto the AST 5-piece
/// recognition lifecycle, sharing a `requestId` across the two roles so the
/// orchestrator can pair source ↔ translation.
public final class AstPolychatService: NativeAstService {

    private static let log = OSLog(subsystem: "com.aiagent.ast_polychat", category: "Service")

    private let session = VoitransWebRtcSession()
    private weak var callback: AstCallback?

    private var currentRequestId: String?
    private var sourceRoleOpen = false
    private var translatedRoleOpen = false

    public init() {}

    deinit {
        session.release()
    }

    public func initialize(configJson: String) {
        guard let data = configJson.data(using: .utf8),
              let cfg = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            os_log("initialize: invalid config json", log: Self.log, type: .error)
            return
        }
        let baseUrl = (cfg["baseUrl"] as? String) ?? ""
        VoitransWebRtcSession.warmupHttp(baseUrl: baseUrl)
        session.initialize(
            baseUrl: baseUrl,
            appId: (cfg["appId"] as? String) ?? "",
            appSecret: (cfg["appSecret"] as? String) ?? "",
            agentId: (cfg["agentId"] as? String) ?? ""
        )
    }

    public func connect(callback: AstCallback) {
        self.callback = callback
        resetRoundState()
        session.connect(handler: VoitransWebRtcSession.EventHandler(
            onConnected: {
                callback.onConnected()
            },
            onMessage: { [weak self] json in
                self?.handleDataChannelMessage(json)
            },
            onDisconnected: { [weak self] in
                self?.forceEndRound()
                callback.onDisconnected()
            },
            onError: { code, message in
                callback.onError(code: code, message: message)
            }
        ))
    }

    public func startAudio() {
        session.startAudio()
    }

    public func stopAudio() {
        session.stopAudio()
    }

    public func interrupt() {
        // WebRTC mode: server-side VAD handles barge-in.
    }

    public func release() {
        forceEndRound()
        session.release()
        callback = nil
    }

    // MARK: - DataChannel event mapping

    private func handleDataChannelMessage(_ json: [String: Any]) {
        guard let cb = callback else { return }
        let type = (json["type"] as? String) ?? ""
        switch type {
        case "user_speaking":
            beginRound(cb, force: true)
            openRole(cb, role: .source)

        case "trans_original":
            beginRound(cb)
            openRole(cb, role: .source)
            let text = (json["text"] as? String) ?? ""
            let done = (json["done"] as? Bool) ?? false
            if !text.isEmpty, let rid = currentRequestId {
                if done {
                    cb.onRecognized(role: .source, requestId: rid, text: text)
                    closeRole(cb, role: .source)
                    maybeEndRound(cb)
                } else {
                    cb.onRecognizing(role: .source, requestId: rid, text: text)
                }
            }

        case "trans_translated":
            beginRound(cb)
            openRole(cb, role: .translated)
            let text = (json["text"] as? String) ?? ""
            let done = (json["done"] as? Bool) ?? false
            if !text.isEmpty, let rid = currentRequestId {
                if done {
                    cb.onRecognized(role: .translated, requestId: rid, text: text)
                    closeRole(cb, role: .translated)
                    maybeEndRound(cb)
                } else {
                    cb.onRecognizing(role: .translated, requestId: rid, text: text)
                }
            }

        case "error":
            let message = (json["message"] as? String) ?? "Unknown error"
            let fatal = (json["fatal"] as? Bool) ?? false
            if fatal {
                cb.onError(code: "ast.fatal", message: message)
            } else {
                cb.onRecognitionError(requestId: currentRequestId, role: nil,
                                       code: "ast.error", message: message)
            }

        default:
            // session_state / mcp_tool_* / bot_response_* — not surfaced
            // through AstCallback.
            break
        }
    }

    // MARK: - Round state machine

    private func beginRound(_ cb: AstCallback, force: Bool = false) {
        if currentRequestId != nil {
            if !force { return }
            if sourceRoleOpen { closeRole(cb, role: .source) }
            if translatedRoleOpen { closeRole(cb, role: .translated) }
            endRound(cb)
        }
        currentRequestId = newRequestId()
    }

    private func openRole(_ cb: AstCallback, role: AstRole) {
        guard let rid = currentRequestId else { return }
        switch role {
        case .source:
            if !sourceRoleOpen {
                sourceRoleOpen = true
                cb.onRecognitionStart(role: role, requestId: rid)
            }
        case .translated:
            if !translatedRoleOpen {
                translatedRoleOpen = true
                cb.onRecognitionStart(role: role, requestId: rid)
            }
        }
    }

    private func closeRole(_ cb: AstCallback, role: AstRole) {
        guard let rid = currentRequestId else { return }
        switch role {
        case .source:
            if sourceRoleOpen {
                sourceRoleOpen = false
                cb.onRecognitionDone(role: role, requestId: rid)
            }
        case .translated:
            if translatedRoleOpen {
                translatedRoleOpen = false
                cb.onRecognitionDone(role: role, requestId: rid)
            }
        }
    }

    private func maybeEndRound(_ cb: AstCallback) {
        if sourceRoleOpen || translatedRoleOpen { return }
        if currentRequestId == nil { return }
        endRound(cb)
    }

    private func endRound(_ cb: AstCallback) {
        guard let rid = currentRequestId else { return }
        cb.onRecognitionEnd(requestId: rid)
        resetRoundState()
    }

    private func forceEndRound() {
        guard let cb = callback else {
            resetRoundState()
            return
        }
        if currentRequestId == nil { return }
        if sourceRoleOpen { closeRole(cb, role: .source) }
        if translatedRoleOpen { closeRole(cb, role: .translated) }
        endRound(cb)
    }

    private func resetRoundState() {
        currentRequestId = nil
        sourceRoleOpen = false
        translatedRoleOpen = false
    }

    private func newRequestId() -> String {
        let ms = Int64(Date().timeIntervalSince1970 * 1000)
        let rand = String(UInt32.random(in: 0..<UInt32(1 << 30)), radix: 36)
        let padded = String(repeating: "0", count: max(0, 6 - rand.count)) + rand
        return "ast_polychat_\(ms)_\(padded)"
    }
}
